src/Symfony/Component/Workflow/Workflow.php

```diff
@@ -363,48 +359,70 @@ private function enter(object $subject, Transition $transition, Marking $marking
-            return;
-        }
+        if ($this->shouldDispatchEvent(WorkflowEvents::ENTERED)) {
+            $event = new EnteredEvent($subject, $marking, $transition, $this);
 
-        $event = new EnteredEvent($subject, $marking, $transition, $this);
+            $this->dispatcher->dispatch($event, WorkflowEvents::ENTERED);
+            $this->dispatcher->dispatch($event, sprintf('workflow.%s.entered', $this->name));
 
-        $this->dispatcher->dispatch($event, WorkflowEvents::ENTERED);
-        $this->dispatcher->dispatch($event, sprintf('workflow.%s.entered', $this->name));
-
-        if ($transition) {
-            foreach ($transition->getTos() as $place) {
-                $this->dispatcher->dispatch($event, sprintf('workflow.%s.entered.%s', $this->name, $place));
+            if ($transition) {
+                foreach ($transition->getTos() as $place) {
+                    $this->dispatcher->dispatch($event, sprintf('workflow.%s.entered.%s', $this->name, $place));
+                }
             }
         }
     }
 
     private function completed(object $subject, Transition $transition, Marking $marking): void
     {
-        if (null === $this->dispatcher) {
-            return;
+        if ($this->shouldDispatchEvent(WorkflowEvents::COMPLETED)) {
+            $event = new CompletedEvent($subject, $marking, $transition, $this);
+
+            $this->dispatcher->dispatch($event, WorkflowEvents::COMPLETED);
+            $this->dispatcher->dispatch($event, sprintf('workflow.%s.completed', $this->name));
+            $this->dispatcher->dispatch($event, sprintf('workflow.%s.completed.%s', $this->name, $transition->getName()));
         }
+    }
 
-        $event = new CompletedEvent($subject, $marking, $transition, $this);
+    private function announce(object $subject, Transition $initialTransition, Marking $marking): void
+    {
+        if ($this->shouldDispatchEvent(WorkflowEvents::ANNOUNCE)) {
+            $event = new AnnounceEvent($subject, $marking, $initialTransition, $this);
+
+            $this->dispatcher->dispatch($event, WorkflowEvents::ANNOUNCE);
+            $this->dispatcher->dispatch($event, sprintf('workflow.%s.announce', $this->name));
 
-        $this->dispatcher->dispatch($event, WorkflowEvents::COMPLETED);
-        $this->dispatcher->dispatch($event, sprintf('workflow.%s.completed', $this->name));
-        $this->dispatcher->dispatch($event, sprintf('workflow.%s.completed.%s', $this->name, $transition->getName()));
+            foreach ($this->getEnabledTransitions($subject) as $transition) {
+                $this->dispatcher->dispatch($event, sprintf('workflow.%s.announce.%s', $this->name, $transition->getName()));
+            }
+        }
     }
 
-    private function announce(object $subject, Transition $initialTransition, Marking $marking): void
+    private function shouldDispatchEvent(string $eventName): bool
     {
+        // If we don't have a dispatcher we can't dispatch the
```

Comment:
I think this comment is a little useless
Also if you use constants for next comparaison, comments may also be useless

Reply:
I agree